marine.c

```diff
@@ -966,6 +977,374 @@ WS_DLL_PUBLIC void marine_free(marine_result *ptr) {
+        }
+        g_array_append_val(indices, i);
+    }
+}
+
+static void
+join_duplicate_fields(marine_packet_field *field, GArray *new_children, GArray *indices, char *key) {
+    // Join the values for duplicate keys, adding the resulting field to new_children.
+    char *new_name = g_strdup(key);
+    marine_packet_field value_child;
+    marine_packet_field_init(&value_child, new_name);
+    value_child.value.type = MARINE_VT_NONE;
+
+    marine_packet_field tree_child;
+    marine_packet_field_init(&tree_child, copy_proto_node_name(new_name, FALSE));
+    tree_child.value.type = MARINE_VT_NONE;
+    tree_child.value.len = 0;
+
+    GArray *joined_values = (GArray *) g_array_new(TRUE, TRUE, sizeof(marine_packet_field_value));
+    GArray *joined_children = (GArray *) g_array_new(TRUE, TRUE, sizeof(marine_packet_field));
+
+    for (unsigned int indices_index = 0; indices_index < indices->len; indices_index++) {
+        int child_index = g_array_index(indices, int, indices_index);
+        marine_packet_field child = g_array_index(field->children, marine_packet_field, child_index);
+        if (child.children == NULL) {
+            g_array_append_val(joined_values, child.value);
+        } else {
+            g_array_append_vals(joined_children, child.children->data, child.children->len);
+            g_array_free(child.children, TRUE); // All children where copied by value to the new array, so we can safely free them.
+        }
+        free(child.name);
+    }
+
+    if (joined_values->len > 0) {
+        value_child.value.type = MARINE_VT_LIST;
+        value_child.value.list_value = (marine_packet_field_value *) joined_values->data;
+        value_child.value.len = joined_values->len;
+        g_array_free(joined_values, FALSE); // Frees only the array metadata, but not the actual values
+        g_array_append_val(new_children, value_child);
+    } else {
+        g_array_free(joined_values, TRUE);
+        marine_packet_field_free(&value_child);
+    }
+
+    if (joined_children->len > 0) {
+        tree_child.children = joined_children;
+        marine_packet_field_join_children_with_duplicate_names(&tree_child);
+        g_array_append_val(new_children, tree_child);
+    } else {
+        g_array_free(joined_children, TRUE);
+        marine_packet_field_free(&tree_child);
+    }
+}
+
+
+/**
+ * Joines children with duplicate names for this packet field.
+ *
+ * Doesn't affect fields with no value.
+ *
+ * The algorithm for this function works as follows:
+ *      1. Map each field name to it's indices in the child array (using a GHashTable)
+ *      2. Create a new children array
+ *      3. Add all the `tree` children (those with children and no value) to the new children, do not join them
+ *      4. Add all value field that appear only once - There is no need to join them
+ *      5. Join the values of all fields with the same name, attach them to a new child and add it to the new children array
+ * @param field
```

Comment:
We don't really use `@param` in our docs here

Reply:
Removed.